examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1138 @@
+    // above is that the global matrix has zero rows and columns for the
+    // algebraic (constrained) components of the solution; the function
+    // puts a value on the diagonal that is nonzero and has about the
+    // same size as the remaining diagonal entries of the matrix. What
+    // this diagonal value is is unknown to us -- in other cases where
+    // we call `current_constraints.distribute_local_to_global()` on
+    // both the left side matrix and the right side vector, as in most
+    // other tutorial programs, the matrix diagonal entries and right
+    // hand side values are chosen in such a way that the result of
+    // solving a linear system is what we want it to be, but the scaling
+    // is done automatically.
+    //
+    // This is not good enough for us here, because we are building the
+    // right hand side independently from the matrix in different functions.
+    // Thus, for any constrained degree of freedom, we set the diagonal of the
+    // Jacobian to be one. This leaves the Jacobian matrix invertible,
+    // consistent with what the time stepper expects, and it also makes sure
+    // that if we did not make a mistake in the residual and/or in the Jacbian
+    // matrix, then asking the time stepper to check the Jacobian with a finite
+    // difference method will produce the correct result. This can be activated
+    // at run time via passing the `-snes_test_jacobian` option on the command
+    // line.
+    for (const auto &c : current_constraints.get_lines())
+      jacobian_matrix.set(c.index, c.index, 1.0);
+    jacobian_matrix.compress(VectorOperation::insert);
+  }
+
+
+  // @sect4{The HeatEquation::solve_with_jacobian() function}
+  //
+  // This is the function that actually solves the linear system with the
+  // Jacobian matrix we have previously built (in a call to the previous
+  // function during the current time step or another earlier one -- time
+  // steppers are quite sophisticated in determining internally whether it is
+  // necessary to update the Jacobian matrix, or whether one can reuse it for
+  // another time step without rebuilding it; this is similar to how one can
+  // re-use the Newton matrix for several Newton steps, see for example the
+  // discussion in step-77). We could in principle not provide this function to
+  // the time stepper, and instead select a specific solver on the command line
+  // by using the `-ksp_*` options of PETSc. However, by providing this
+  // function, we can use a specific solver and preconditioner for the linear
+  // system, and still have the possibility to change them on the command line.
+  //
+  // Providing a specific solver is more in line with the way we usually do
+  // things in other deal.II examples, while letting PETSc choose a generic
+  // solver, and changing it on the command line via `-ksp_type` is more in line
+  // with the way PETSc is usually used. Both options are available here, since
```

Comment:
```suggestion
  // with the way PETSc is usually used, and it can be a convenient approach when we are experimenting to find an optimal solver for our problem. Both options are available here, since
```